File: src/recursion/java/FamilyTreeTest.java
package recursion.java;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.*;
public class FamilyTreeTest {
    private FamilyTree familyTree;
    private FamilyTree child1;

    @Before
    public void setUp() {
        familyTree = new FamilyTree("John", 80);
        child1 = new FamilyTree("Jane", 50);
        FamilyTree child2 = new FamilyTree("Jack", 45);
        FamilyTree grandchild1 = new FamilyTree("Jill", 25);
        FamilyTree grandchild2 = new FamilyTree("Joe", 20);
        FamilyTree grandchild3 = new FamilyTree("Jenny", 18);
        FamilyTree grandchild4 = new FamilyTree("Jerry", 15);
        child1.addChild(grandchild1);
        child2.addChild(grandchild2);
        child2.addChild(grandchild3);
        child2.addChild(grandchild4);
        familyTree.addChild(child1);
        familyTree.addChild(child2);
    }

    @After
    public void tearDown() {

    }

    @Test
    public void calculateTotalAgeRecursive() {
        assertEquals("Test1 failed. Expected 253, but got " + familyTree.calculateTotalAgeRecursive(),
                253, familyTree.calculateTotalAgeRecursive());
        System.out.println("Test1 passed.");
        assertEquals("Test2 failed. Expected 95, but got " + child1.calculateTotalAgeRecursive(),
                95, child1.calculateTotalAgeRecursive());
        System.out.println("Test2 passed.");
    }

    @Test
    public void findYoungestMember() {
        assertEquals("Test1 failed. Expected Joe, but got " + familyTree.findYoungestMember().getName(),
                "Joe", familyTree.findYoungestMember().getName());
        System.out.println("Test1 passed.");
        assertEquals("Test2 failed. Expected Jill, but got " + child1.findYoungestMember().getName(),
                "Jill", child1.findYoungestMember().getName());
        System.out.println("Test2 passed.");
    }

    @Test
    public void findShortestFamilyLine() {
        FamilyTree greatgrandfather = new FamilyTree("Grandpa", 100);
        FamilyTree granduncle = new FamilyTree("Granduncle", 80);
        greatgrandfather.addChild(granduncle);
        greatgrandfather.addChild(familyTree);
        assertEquals("Test1 failed. Expected [Grandpa, Granduncle], but got " + familyTree.findShortestFamilyLine().toString(),
                "[Grandpa, Granduncle]", familyTree.findShortestFamilyLine().toString());
        System.out.println("Test1 passed.");

    }

    @Test
    public void findFamilyWithMostChildren() {
        assertEquals("Test1 failed. Expected name to be Jack, got: " +
                familyTree.findFamilyWithMostChildren().getName(), "Jack", familyTree.findFamilyWithMostChildren().getName());
        System.out.println("Test1 passed.");
    }
}

File: src/recursion/python/FamilyTreeTest.py
import unittest
from unittest import TestCase
from src.recursion.python.FamilyTree import FamilyTree

class FamilyTreeTest(TestCase):
    def setUp(self):
        self.family_tree = FamilyTree("John", 80)
        self.child1 = FamilyTree("Jane", 50)
        child2 = FamilyTree("Jack", 45)
        grandchild1 = FamilyTree("Jill", 25)
        grandchild2 = FamilyTree("Joe", 20)
        grandchild3 = FamilyTree("Jenny", 18)
        grandchild4 = FamilyTree("Jerry", 15)
        self.child1.add_child(grandchild1)
        child2.add_child(grandchild2)
        child2.add_child(grandchild3)
        child2.add_child(grandchild4)
        self.family_tree.add_child(self.child1)
        self.family_tree.add_child(child2)

    def test_find_youngest_member(self):
        self.assertEqual(self.family_tree.find_youngest_member().get_name(), "Jerry")
        self.assertEqual(self.child1.find_youngest_member().get_name(), "Jill")

    def test_find_shortest_family_line(self):
        greatgrandfather = FamilyTree("Grandpa", 100)
        granduncle = FamilyTree("Granduncle", 80)
        greatgrandfather.add_child(granduncle)
        greatgrandfather.add_child(self.family_tree)
        self.assertEqual(str(greatgrandfather.find_shortest_family_line()), "['Grandpa', 'Granduncle']")

    def test_find_family_with_most_children(self):
        self.assertEqual(self.family_tree.find_family_with_most_children().get_name(), "Jack")

if __name__ == '__main__':
    unittest.main()